### ISSUE TITLE
Deprecation warnings emitted in crypto_aes.c when building against OpenSSL 3.0

### ISSUE THREAD

**Issue by TimWolla** [Tarsnap/libcperciva#465]
This was noticed when updating the spiped Docker image to Alpine 3.17 which ships with OpenSSL 3.0:

https://github.com/TimWolla/docker-spiped/actions/runs/3594642501/jobs/6053158056

```
  gcc  -D_POSIX_C_SOURCE=200809L -D_XOPEN_SOURCE=700 -DCPUSUPPORT_CONFIG_FILE=\"cpusupport-config.h\"  -I.. -I../libcperciva/alg -I../libcperciva/cpusupport -I../libcperciva/crypto -I../libcperciva/datastruct -I../libcperciva/events -I../libcperciva/netbuf -I../libcperciva/network -I../libcperciva/util -I../lib/dnsthread -I../lib/proto -I../lib/util  -O1 -c ../libcperciva/crypto/crypto_aes.c -o crypto_aes.o
  ../libcperciva/crypto/crypto_aes.c: In function 'openssl_oneshot':
  ../libcperciva/crypto/crypto_aes.c:150:9: warning: 'AES_set_encrypt_key' is deprecated: Since OpenSSL 3.0 [-Wdeprecated-declarations]
    150 |         AES_set_encrypt_key(key, (int)(len * 8), &kexp);
        |         ^~~~~~~~~~~~~~~~~~~
  In file included from ../libcperciva/crypto/crypto_aes.c:6:
  /usr/include/openssl/aes.h:51:5: note: declared here
     51 | int AES_set_encrypt_key(const unsigned char *userKey, const int bits,
        |     ^~~~~~~~~~~~~~~~~~~
  ../libcperciva/crypto/crypto_aes.c:151:9: warning: 'AES_encrypt' is deprecated: Since OpenSSL 3.0 [-Wdeprecated-declarations]
    151 |         AES_encrypt(ptext, ctext, &kexp);
        |         ^~~~~~~~~~~
  /usr/include/openssl/aes.h:57:6: note: declared here
     57 | void AES_encrypt(const unsigned char *in, unsigned char *out,
        |      ^~~~~~~~~~~
  ../libcperciva/crypto/crypto_aes.c: In function 'crypto_aes_key_expand':
  ../libcperciva/crypto/crypto_aes.c:249:9: warning: 'AES_set_encrypt_key' is deprecated: Since OpenSSL 3.0 [-Wdeprecated-declarations]
    249 |         AES_set_encrypt_key(key, (int)(len * 8), kexp);
        |         ^~~~~~~~~~~~~~~~~~~
  /usr/include/openssl/aes.h:51:5: note: declared here
     51 | int AES_set_encrypt_key(const unsigned char *userKey, const int bits,
        |     ^~~~~~~~~~~~~~~~~~~
  ../libcperciva/crypto/crypto_aes.c: In function 'crypto_aes_encrypt_block':
  ../libcperciva/crypto/crypto_aes.c:285:9: warning: 'AES_encrypt' is deprecated: Since OpenSSL 3.0 [-Wdeprecated-declarations]
    285 |         AES_encrypt(in, out, (const void *)key);
        |         ^~~~~~~~~~~
  /usr/include/openssl/aes.h:57:6: note: declared here
     57 | void AES_encrypt(const unsigned char *in, unsigned char *out,
        |      ^~~~~~~~~~~
```

The resulting binary appears to be fully functional.

**Comment by gperciva** [comment]
Thanks for the report!

Yes, it's unfortunate, but is harmless for now.  OpenSSL is deprecating their "low-level" APIs, such as `AES_encrypt`:

> Use of the low level APIs has been informally discouraged by the OpenSSL development team for a long time. However in OpenSSL 3.0 this is made more formal. All such low level APIs have been deprecated. You may still use them in your applications, but you may start to see deprecation warnings during compilation (dependent on compiler support for this). Deprecated APIs may be removed from future versions of OpenSSL so you are strongly encouraged to update your code to use the high level APIs instead.

https://www.openssl.org/docs/man3.0/man7/migration_guide.html#Low-Level-APIs

Unfortunately, we can't switch to their "high level APIs", so we're looking at alternate solutions.

Fortunately, any modern x86 machine (and some modern ARM machines) will use our hardware-accelerated intrinsics, so in theory we could simply omit the openssl parts from compilation.  (That works perfectly well in a test branch of mine.)

**Comment by TimWolla** [comment]
> Yes, it's unfortunate, but is harmless for now. 

Yeah, my primary intention with this issue was to give you sufficient heads-up time.

> in theory we could simply omit the openssl parts from compilation. (That works perfectly well in a test branch of mine.)

I believe this won't work if the binary is compiled on one CPU (with support) and executed on another one (without), no? This would be the case for Docker, you just grab the pre-built Docker image and are up and running. The same applies to any binary distro (e.g. Debian).

**Comment by gperciva** [comment]
> I believe this won't work if the binary is compiled on one CPU (with support) and executed on another one (without), no?

Yes, that's true.  :(

It's quite unfortunate that OpenSSL deprecating the public API for `AES_encrypt`.  It would be technically possible to include the relevant source files directly in our repository, but that has legal issues.  Or we could find an alternate AES implementation, or else code it ourselves.  All of those options would require us to maintain that code, though.

If OpenSSL was completely removing some code, I'd say "cool beans" and move on with life.  But they're obviously going to be keeping AES around for their own internal use, so it's rather unfortunate that we're going to have to maintain a separate AES implementation purely because of them deprecating it in their public API.

**Comment by cperciva** [comment]
Just to be clear, it doesn't matter which instruction set extensions are supported by the CPU we're compiling on -- only which instruction set extensions are supported by the *compiler* we're using.  We detect at runtime whether we should use the AESNI, SHANI, RDRAND, SSE2, etc variants of routines, and compile in everything the compiler can handle.

**Comment by gperciva** [comment]
@cperciva is correct about the current state of libcperciva.

For clarification, when I said:

> in theory we could simply omit the openssl parts from compilation

I was talking about #399, which he has already declined.  I think the 2 comments in that PR accurately sum up our positions.
